plutus-tx-plugin/executables/profile/Main.hs

```diff
@@ -24,12 +41,117 @@ import qualified PlutusCore.Default        as PLC
 
 import           Control.Lens.Combinators  (_2)
 import           Control.Lens.Getter       (view)
+import           Data.List                 (intercalate)
+import           Data.Maybe                (fromJust)
 import           Data.Proxy                (Proxy (Proxy))
 import           Data.Text                 (Text)
+import           Data.Time.Clock           (NominalDiffTime, UTCTime, diffUTCTime)
 import           Prettyprinter.Internal    (pretty)
 import           Prettyprinter.Render.Text (hPutDoc)
 import           System.IO                 (IOMode (WriteMode), withFile)
 
+data Stacks
+  = MkStacks
+  { -- | The variable name.
+    varName           :: String,
+    -- | The time when it starts to be evaluated.
+    startTime         :: UTCTime,
+    -- | The time spent on evaluating the functions it called.
+    timeSpentCalledFn :: NominalDiffTime
+  }
+  deriving (Show)
+
+-- | Write the time log of a program to a file in
+-- the plutus-tx-plugin/executables/profile/ directory.
+writeLogToFile ::
+  ToUPlc a PLC.DefaultUni PLC.DefaultFun =>
+  -- | Name of the file you want to save it as.
+  FilePath ->
+  -- | The program to be profiled.
+  [a] ->
+  IO ()
+writeLogToFile fileName values = do
+  let filePath = "plutus-tx-plugin/executables/profile/"<>fileName<>".timelog"
+  log <- pretty . view _2 <$> (rethrow $ runUPlcProfileExec values)
+  withFile
+    filePath
+    WriteMode
+    (\h -> hPutDoc h log)
+  processed <- processLog filePath
+  writeFile (filePath<>".stacks") processed
+  pure ()
+
+processLog :: FilePath -> IO [Char]
+processLog file = do
+  content <- readFile file
+  -- lEvents is in the form of [[t1,t2,t3,entering/exiting,var]]. Time is chopped to 3 parts.
+  let lEvents =
+        map
+          -- @tail@ strips "[" in the first line and "," in the other lines,
+          -- @words@ turns it to a list of [t1,t2,t3, enter/exit, var]
+          (tail . words)
+          -- turn to a list of events
+          (lines content)
+      lTime = map (unwords . take 3) lEvents
+      -- list of enter/exit
+      lEnterOrExit = map (!! 3) lEvents
+      -- list of var
+      lVar = map (!! 4) lEvents
+      lTripleTimeVar = zip3 (lUTC lTime) lEnterOrExit lVar
+      stacks = getStacks [] lTripleTimeVar
+      fnsStacks = map (intercalate "; " . fst) stacks
+      stacksFgFormat (hdf:tlf) (hdt:tlt)=
+        hdf<>" "<>show hdt<>"\n":stacksFgFormat tlf tlt
+      stacksFgFormat _ _ = []
+  pure $
+    concat $
```

Comment:
I'd just return a list from this function. Or, tbh, a `[StackTime]` or whatever, and let some other function print it.